plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/VectorFunctions.java

```diff
@@ -0,0 +1,110 @@
+import static io.trino.spi.type.RealType.REAL;
+import static io.trino.spi.type.StandardTypes.DOUBLE;
+
+public final class VectorFunctions
+{
+    private VectorFunctions() {}
+
+    @Description("Calculates the euclidean distance between two vectors. This is same as L2 distance <-> in pgvector.")
+    @ScalarFunction
+    @SqlType(DOUBLE)
+    public static double euclideanDistance(@SqlType("array(real)") Block first, @SqlType("array(real)") Block second)
+    {
+        checkArgument(first.getPositionCount() == second.getPositionCount(), "The arguments must have the same length");
+
+        double sum = 0.0;
+        for (int i = 0; i < first.getPositionCount(); i++) {
+            double firstValue = REAL.getFloat(first, i);
+            double secondValue = REAL.getFloat(second, i);
+            sum = sum + Math.pow((firstValue - secondValue), 2.0);
+        }
+        return Math.sqrt(sum);
+    }
+
+    @Description("Calculates the inner product between two vectors. This is same as cosine distance <#> in pgvector.")
+    @ScalarFunction
+    @SqlType(DOUBLE)
+    public static double innerProduct(@SqlType("array(real)") Block first, @SqlType("array(real)") Block second)
+    {
+        checkArgument(first.getPositionCount() == second.getPositionCount(), "The arguments must have the same length");
+
+        double[] firstVector = new double[first.getPositionCount()];
+        double[] secondVector = new double[second.getPositionCount()];
+
+        for (int i = 0; i < first.getPositionCount(); i++) {
+            firstVector[i] = REAL.getFloat(first, i);
+            secondVector[i] = REAL.getFloat(second, i);
+        }
+        return -dotProduct(firstVector, secondVector);
+    }
+
+    @Description("Calculates the cosine distance between two vectors. This is same as cosine distance <=> in pgvector.")
+    @ScalarFunction
+    @SqlType(DOUBLE)
+    public static double cosineDistance(@SqlType("array(real)") Block first, @SqlType("array(real)") Block second)
+    {
+        checkArgument(first.getPositionCount() == second.getPositionCount(), "The arguments must have the same length");
+
+        double[] firstVector = new double[first.getPositionCount()];
+        double[] secondVector = new double[second.getPositionCount()];
+
+        for (int i = 0; i < first.getPositionCount(); i++) {
+            firstVector[i] = REAL.getFloat(first, i);
+            secondVector[i] = REAL.getFloat(second, i);
+        }
+        double dotProduct = dotProduct(firstVector, secondVector);
+        double firstMagnitude = magnitude(firstVector);
+        double secondMagnitude = magnitude(secondVector);
+        checkArgument(firstMagnitude != 0 && secondMagnitude != 0, "Vector magnitude cannot be zero");
+
+        double cosineSimilarity = dotProduct / (firstMagnitude * secondMagnitude);
+        return 1.0 - cosineSimilarity;
+    }
+
+    private static double dotProduct(double[] first, double[] second)
```

Comment:
It looks like the callers all have floats, so this could use floats for the argument and then have a double for the computation.  This would keep temp memory down.

Also, as @raunaqmorarka mentioned we could likely just work directly with the blocks, or pull out the underlying arrays (see `IntArrayBlock` `getRawValues` and `getRawValuesOffset`).  For the moment these are int[], but I expect we will move to either `float[]` or `MemorySegment` soon.